### PR TITLE
[Data] [Docs] Additional documentation for `BatchPredictor` deprecation

### PR DESCRIPTION
**Pull request by amogkam**
1. Updates the deprecation message to link to the corresponding issue.
2. Adds documentation in Ray Data docs for batch inference with models trained with Ray Train.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/batch_inference.rst [hunk 1]
```diff
@@ -418,3 +418,92 @@ Suppose your cluster has 4 nodes, each with 16 CPUs. To limit to at most
...
+    test_dataset = valid_dataset.drop_columns(["target"])
+
+    class XGBoostPredictor:
+        def __init__(self, checkpoint: Checkpoint):
+            xgboost_checkpoint = XGBoostCheckpoint.from_checkpoint(checkpoint)
+            self.model = xgboost_checkpoint.get_model()
+        
+        def __call__(self, data: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+            # Stack all the features into a single Numpy array.
+            features = np.stack(list(data.values()), axis=1)
```
Comment (pcmoritz) [inline doc/source/data/batch_inference.rst:489]: This depends on the ordering of the dict -- how do you make sure the ordering was the same during training?

File: doc/source/data/batch_inference.rst [hunk 2]
```diff
@@ -418,3 +418,92 @@ Suppose your cluster has 4 nodes, each with 16 CPUs. To limit to at most
...
+    # Use 2 parallel actors for inference. Each actor predicts on a
+    # different partition of data.
+    scale = ray.data.ActorPoolStrategy(size=2)
+    # Map the Predictor over the Dataset to get predictions.
+    predictions = test_dataset.map_batches(
+        XGBoostPredictor, 
+        compute=scale,
+        # Pass in the Checkpoint to the XGBoostPredictor constructor.
+        fn_constructor_kwargs={"checkpoint": checkpoint}
+        )
```
Comment (pcmoritz) [inline doc/source/data/batch_inference.rst:503]: The parenthesis should be unindented by 4 spaces :)

File: doc/source/data/batch_inference.rst (hunk 1 above)
Comment (amogkam) [reply doc/source/data/batch_inference.rst:489]: If this didn't hold, then there would lots of issues with Ray Data right now :)

**Review by bveeramani — APPROVED** [review]

File: doc/source/data/batch_inference.rst (hunk 1 above)
Comment (pcmoritz) [reply doc/source/data/batch_inference.rst:489]: It may very well be the case, but it is very dangerous to depend on such things. We should always design things in a way that makes models not depend on implementation details like this. First, do you really understand why the dict insertion order is guaranteed? Will it not break in the future? What should a user do if they want to construct their own batch for prediction, they won't know the right insertion order, right?

Models are artifacts that should not depend on implementation details of the code. Can we improve this?

Comment (amogkam) [reply doc/source/data/batch_inference.rst:489]: This is a core primitive that Ray Datasets relies upon (and is guaranteed by Python). If you are saying we cannot rely on this, then Ray Datasets as a whole is not robust, and all of our examples are incorrect.

Comment (pcmoritz) [reply doc/source/data/batch_inference.rst:489]: So you are saying the ordering of the keys in the batch dict is part of the API contract? Is this documented somewhere (if not, we should add it)? Do we say where the canonical ordering is coming from (e.g. how it is mapping from the original datasource) so we can give users guidance on how they should construct the dicts if they do it manually without Ray Data.

This is very non-obvious I think btw :)

Comment (amogkam) [reply doc/source/data/batch_inference.rst:489]: Basically, dict ordering is guaranteed by Python. Column ordering is also guaranteed by Pyarrow. Ray Datasets depends on Python and Pyarrow. By definition, we assume that Python and Pyarrow work as intended.

Comment (pcmoritz) [reply doc/source/data/batch_inference.rst:489]: Alright, let's call this out in the docs then. We should say that the Ray Data dicts for the batches will always have the same ordering as the columns in the underlying files. Let's also make sure we go through all the possible ways how datasets can be constructed and that this is actually something we can guarantee :)

This is important, because if we do things like in this example, we need to make sure somebody can use the model without Ray Data :)

Comment (amogkam) [reply doc/source/data/batch_inference.rst:489]: Hmm maybe I'm misunderstanding you. Are you talking about the case where you don't use Ray Data for training, but you are using it for inference?

In this particular example, both the train set and test set are coming from the same original dataset. So therefore, the column ordering is the same all throughout.

Comment (pcmoritz) [reply doc/source/data/batch_inference.rst:489]: I'm talking about the case where you use Ray Data for the training, and then later you load the checkpoint and do the prediction with your own custom predictor class. In that case you need to know what the ordering or the features is you need to pass in :)

Much like in your example but instead of doing batch prediction with Ray Data, you just do prediction on your own, e.g. on a single example.

Comment (pcmoritz) [reply doc/source/data/batch_inference.rst:489]: To be more concrete, in this example it would be
```
predictor = XGBoostPredictor(checkpoint)
predictor.predict({"first_feature": 1, "second_feature": 2, ...})
```
People need to know that if they put the features in the right order in the dict, things will work, and that we guarantee that.

Comment (amogkam) [reply doc/source/data/batch_inference.rst:489]: Makes sense, though I'm not sure what needs to be changed here? We don't use `XGBoostPredictor` in the example.

Comment (amogkam) [reply doc/source/data/batch_inference.rst]: ok ok, I get what you're saying now. You're not talking about the Python implementation details, but the implementation details on how the arrow tables are inserted into the numpy dict.

I switched to pandas batch format to avoid any confusion. Generally, for this type of tabular data, pandas is probably the better option anyways.

Comment (pcmoritz) [reply doc/source/data/batch_inference.rst]: Yeah, that's what I meant :)

Sounds good, I think you are right pandas is clearer here. For serving probably not though -- you wouldn't want pandas on the critical path of your serving since it can have many conversion surprises with unclean data.

**Review by pcmoritz — APPROVED** [review]
